MaterialLib/SolidModels/Ehlers-impl.h

```diff
@@ -740,6 +740,94 @@ SolidEhlers<DisplacementDim>::integrateStress(
         tangentStiffness)};
 }
 
+template <int DisplacementDim>
+std::vector<typename MechanicsBase<DisplacementDim>::InternalVariable>
+SolidEhlers<DisplacementDim>::getInternalVariables() const
+{
+    return {
+        {"damage.kappa_d", 1,
+         [](typename MechanicsBase<
+                DisplacementDim>::MaterialStateVariables const& state,
+            std::vector<double>& cache) -> std::vector<double> const& {
+             assert(dynamic_cast<StateVariables<DisplacementDim> const*>(
+                        &state) != nullptr);
+             auto const& ehlers_state =
+                 static_cast<StateVariables<DisplacementDim> const&>(state);
+
+             cache.resize(1);
+             cache.front() = ehlers_state.damage.kappa_d();
+             return cache;
+         }},
+        {"damage.value", 1,
+         [](typename MechanicsBase<
+                DisplacementDim>::MaterialStateVariables const& state,
+            std::vector<double>& cache) -> std::vector<double> const& {
+             assert(dynamic_cast<StateVariables<DisplacementDim> const*>(
+                        &state) != nullptr);
+             auto const& ehlers_state =
+                 static_cast<StateVariables<DisplacementDim> const&>(state);
+
+             cache.resize(1);
+             cache.front() = ehlers_state.damage.value();
+             return cache;
+         }},
+        {"eps_p.D", KelvinVector::RowsAtCompileTime,
+         [](typename MechanicsBase<
+                DisplacementDim>::MaterialStateVariables const& state,
+            std::vector<double>& cache) -> std::vector<double> const& {
+             assert(dynamic_cast<StateVariables<DisplacementDim> const*>(
+                        &state) != nullptr);
+             auto const& ehlers_state =
+                 static_cast<StateVariables<DisplacementDim> const&>(state);
+             auto const& D = ehlers_state.eps_p.D;
+
+             cache.resize(KelvinVector::RowsAtCompileTime);
```

Comment:
AFAIK resize will allocate new memory only if the new size exceeds the old vector capacity. So there won't be a new allocation for every integration point.
But for every element there will be a new `cache` instance (if I remember correctly), and subsequently there will be (at least) one heap allocation per element.
So this approach is not very efficient, but the problem does not in this line, rather it lies in the fact that there is a new cache instance for each element (the advantage of this approach is that there is no "global" cache, which makes the code a bit simpler).
Additionally this is used for output only, so it should not be too expensive.